feat: Storefront Components (#15651)
shopware/platform: src/Storefront/Resources/views/components/README.md

## Anonymous components

* It is recommended to use anonymous components without a PHP class that are declared in twig.
* Components with a PHP class can only be used in plugins.
* Apps only support anonymous components.

## Naming and directory structure

* Each component name must be unique.
* Default storefront components are in the `Sw` namespace. `<twig:Sw:Button>`
* Component names are written uppercase.
* 3rd party components bring their own namespace e.g. `<twig:Agency:Button>`.
* (S)CSS and JavaScript are in the same directory as the component template.

```
views/components/Sw/
    Button.html.twig
    Button.js
    Button.scss
```

You can also use the index naming for anonymous components, which enables you to use the directory name as the component name. This can be helpful to have all component files in a separated directory, but without adding a duplicated namespace level. Make sure to name your JS and SCSS files accordingly.

```
views/components/Sw/Button
    index.html.twig
    index.js
    index.scss
```

Both of the above structures produces the component `Sw:Button`.

✅ Do:
```html
<twig:Sw:Button>My button</twig:Sw:Button>
```

> ❌ Don't

Reviewer: this looks a bit strange 🤔 shouldn't the "don't" be on the directory or file names?

```html
<twig:sw-button>My button</twig:sw-button>
```

## Using a PHP class for the component

> Besides the anonymous components, which only require a Twig template, you can also define your component via a PHP class which offers the possiblity to add additional businsess logic to your component. You can have a look at the [offical documentation](https://symfony.com/bundles/ux-twig-component/current/index.html) for all the details.

Reviewer: ```suggestion
Besides the anonymous components, which only require a Twig template, you can also define your component via a PHP class which offers the possibility to add additional business logic to your component. You can have a look at the [official documentation](https://symfony.com/bundles/ux-twig-component/current/index.html) for all the details.
```

> In Shopware we decided that these PHP classes should be placed right where your component template and other files of your component are located. This provides the epxerience of a real comopnent system and you have all component related files in one place. Therefore you can simply add the PHP class to the described directory structure.

Reviewer: ```suggestion
In Shopware we decided that these PHP classes should be placed right where your component template and other files of your component are located. This provides the experience of a real component system and you have all component related files in one place. Therefore you can simply add the PHP class to the described directory structure.
```

**Important Note:** If you use a PHP class for your component, you cannot use the anonymous component naming feature using `index.html.twig` for your component. The PHP class and the component template should have matching names.

```
views/components/Button/
    Button.html.twig
    Button.js
    Button.php
    Button.scss
```

The loading and template matching is already solved by placing the file in the right directory,
so you don't have to define a specific name or template path in your component class.

```PHP
<?php declare(strict_types=1);

namespace Shopware\Storefront\Resources\views\components\Button;

use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;

#[AsTwigComponent()]
class Button
{
    public string $text = 'Click me!';
}
```

The only thing you have to do though is to register your component class as a service in the service.xml configuration.

> ```XML

Reviewer: as we will soon convert the service declaration to PHP, we might want to document the PHP way already here

Author: I would keep it for now until we do the real migration. The documentation needs to be updated in many other places as well if we change this. But it should be part of that change.

<!-- services.xml -->

<!-- other config ... -->

<service id="Shopware\Storefront\Resources\views\components\Button">
    <tag name="twig.component" />
</service>
```

Now your component is all set up and ready to go. You can directly access it in the template under the given name and namespace.

> **Note:** Be careful with adding too much business logic inside component classes. At Shopware we mostly implement anonymous components that are pure frontend representaions which get passed all necessary data via properties.

Reviewer: ```suggestion
**Note:** Be careful with adding too much business logic inside component classes. At Shopware we typically implement anonymous components that are pure frontend representations which get passed all necessary data via properties.
```

## Adding components via Shopware extension

New components can easily be added by Shopware apps and plugins.

### Anonymous Components

> The anonymous components work right of the box and you dont't have to do anything else than putting a template file under `<your-plugin>/src/Resources/views/components`. Both, apps and plugins can provide anonymous components in any distribution form of Shopware. Your components will automatically be prefixed with the bundle name of your app or plugin.

Reviewer: ```suggestion
The anonymous components work right of the box and you don't have to do anything else than putting a template file under `<your-plugin>/src/Resources/views/components`. Both, apps and plugins can provide anonymous components in any distribution form of Shopware. Your components will automatically be prefixed with the bundle name of your app or plugin.
```

          Button/
            Primary.html.twig

```

This will result in the following component being registered:

```Twig
<twig:MyExtension:Button:Primary />
```

For further information about anonymous components you can have a look at the [official documentation](https://symfony.com/bundles/ux-twig-component/current/index.html#anonymous-components).

### Components with PHP class
Components that make use of a PHP class can only be provided by plugins, not via apps. If you want to use a PHP class for a component in your plugin, you can simply create it next to your other component files, like described above. You can then define your component. Make sure to use the right namespace.

**Example:**

```
MyPlugin/
  src/
    Resources/
      views/
        commponents/
          Button/
            Primary.html.twig
            Primary.php

```

```PHP
<?php declare(strict_types=1);

namespace MyPlugin\Resources\views\components\Button;

use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;

#[AsTwigComponent()]
class Primary
{
    public string $text = 'Click me!';

    // Add more public props and logic here.
}
```

```Twig
<twig:MyPlugin:Button:Primary />
```

> To learn what kind of possibilities the PHP implementation of your component offers, you can just refer to the [offical documentation](https://symfony.com/bundles/ux-twig-component/current/index.html).

Reviewer: ```suggestion
To learn what kind of possibilities the PHP implementation of your component offers, you can just refer to the [official documentation](https://symfony.com/bundles/ux-twig-component/current/index.html).
```

## Best Practices

### Class naming

* Each component must have a unique root class that is used throughout the component markup.

> * The root class uses the `sw-` prefix to be independent of existing/old components.

Reviewer: but `sw-` prefix is only for our own components, right? 🤔

Author: Yes

Reviewer: do we need to explicitly document that?

Author: I think it is good to know. Especially for AI agents to also know the best practices.

* Child elements must "follow" the root class naming.
* The BEM naming pattern is used.

✅ Do:
```html
<div class="sw-product-card card">
    <div class="sw-product-card__body card-body">
        <h2 class="sw-product-card__title">Card title</h2>
    </div>
    <div class="sw-product-card__footer card-footer">
    </div>
</div>
```

```css
.sw-product-card {
    /* Styling */
}

.sw-product-card__body {
    /* Styling */
}

.sw-product-card__title {
    /* Styling */
}
```

❌ Don't
```html
<div class="sw-product-card card">
    <div class="sw-card-inner card-body">
        <h2>Card title</h2>
    </div>
    <div class="sw-card-bottom card-footer">
    </div>
</div>
```

```css
.sw-product-card {
    .sw-card-inner {
        h2 {
            /* Styling */
        }
    }
}
```

### Component APIs and extensibility

A UX component can offer several APIs to allow extension. Extension means that the APIs of a component can be used to configure or modify a component during its usage.

> The APIs of a component are "Props", "Blocks", "Attributes" and "Slots". Slots are an additional feature in shopware to allow component extension using the CMS.

Reviewer: ```suggestion
The APIs of a component are "Props", "Blocks", "Attributes" and "Slots". Slots are an additional feature in Shopware to allow component extension using the CMS.
```

        {# Entry-point for after the core contents of the header. #}
        <twig:Slot name="header-top"></twig:Slot>
    </div>

    {# ... #}
</div>
```

#### Slots with default content:
```twig
{% props
    product,
    slots = [] {# Prop that contains the content of the slots. #}
%}

<div class="sw-product-card card">
    <div class="sw-product-card__header">
        {# Entry-point for before the core contents of the header. #}
        <twig:Slot name="header-top"></twig:Slot>

        {# Entry-point for media with default content. #}
        <tiwg:Slot name="media">
            <img src="..." class="w-product-card__media card-img-top" alt="...">
        </twig:Slot>

        {# Entry-point for after the core contents of the header. #}
        <twig:Slot name="header-top"></twig:Slot>
    </div>

    {# ... #}
</div>
```

#### Slots in combination with blocks:
```twig
{% props
    product,
    slots = [] {# Prop that contains the content of the slots. #}
%}

<div class="sw-product-card card">
    <div class="sw-product-card__header">

        {# Block to overwrite the content programmatically when using the component in a twig template. #}

>         {% block media %}

Reviewer: maybe I got that wrong, but don't we want to prevent blocks in the components? 🤔

Author: We still use blocks in the components, but they work a bit differently. They are the extension points for developers who want to build their custom version of the component.

I explained this in more detail in the discussion.

https://github.com/shopware/shopware/discussions/13806#discussioncomment-15999320

            {# Entry-point for API and CMS. #}
            <tiwg:Slot name="media">
                <img src="..." class="w-product-card__media card-img-top" alt="...">
            </twig:Slot>
        {% endblock %}

    </div>

    {# ... #}
</div>
```

> ### Attributes and CVA

Reviewer: I needed to google again what `CVA` stands for 🙈

Reviewer: what I wanted to say here, do we need to write once, for what this abbreviation stands? 🙂

> * A component should make attributes extendable using the attributes feature of symfony UX.

Reviewer: ```suggestion
* A component should make attributes extendable using the attributes feature of Symfony UX.
```

* Use nested attributes for child-elements.
* A component should also use CVA to make the CSS-classes configurable.
* The CVA must bring the `base` variant by default.
* A prop should be available to allow extending the CVA variants.
* A prop should be available to allow changing of the base classes of the component.

✅ Do:
```twig
{% props
    defaultBaseClasses = 'sw-product-card card',
    defaultVariants = {},
%}
{% set rootVariants = defaultVariants|merge({
    base: defaultBaseClasses,
}) %}
{% set rootCVA = cva(rootVariants) %}
<div {{ attributes.defaults({ role: 'article' }) }} class="{{ rootCVA.apply({}, attributes.render('class')) }}">
    ...
</div>
```

❌ Don't
```twig
<div class="sw-product-card card" role="article">
    ...
</div>
```

### Data independence and global state access

* A component should be as independent as possible.
* A component must not rely on its parent component in order to function correctly.
* A component must not rely on global variables or state internally in order to function correctly.
* If a global setting is needed e.g. `config('core.listing.allowBuyInListing')` it should be able to be passed as a prop from outside.

> * A component can use symfony translation internally.

Reviewer: ```suggestion
* A component can use Symfony translation internally.
```

    /* Uses a SCSS variable when a CSS variable can be used instead. */
    border: 1px solid $border-color;

    .sw-product-card__media {
        /* Hardcodes values that might be worth customizing */
        height: 200px;
    }

    .card-body {
        /* Uses custom selector for hard styling overwrite to customize the padding */
        /* instead of using the Bootstrap CSS variables --bs-card-spacer. */
        padding: 1rem;

        /* Uses deep nesting that is not needed with proper naming */
        .product-card__price {
            font-variant-numeric: tabular-nums;
        }
    }
}
```

## JavaScript

You can add a corresponding JS file for your component. This file is automatically loaded within the template if your component is used on the site. This JS will be integrated as it is, without a bundling process or similar. You can use it for any form of JavaScript to add frontend business logic to your component. It can be plain JS or you can make use of a new component system that we added as an alternative to the current JS plugin system, which can help you create simple encapsulated logic for your component.

> For more information you can have a look at the README.md file in `app/storefron/src/component-system/`.

Reviewer: ```suggestion
For more information you can have a look at the [README.md](../../app/storefront/src/component-system/README.md).
```
